src/game/Movement/WaypointMovementGenerator.cpp

```diff
@@ -181,64 +203,111 @@ bool WaypointMovementGenerator<Creature>::OnArrived(Creature& creature)
 
-    WaypointNode const& nextNode = currPoint->second;
     Movement::MoveSplineInit init(creature, "WaypointMovementGenerator<Creature>::StartMove");
-
-    if (WaypointPath const* pSubpath = (nextNode.path_id ? sWaypointMgr.GetPathFromOrigin(nextNode.path_id, 0, 0, PATH_FROM_SPECIAL) : nullptr))
+    PointsArray genPath; // will contain the generated path
+    genPath.reserve(20); // little optimization
+    genPath.push_back(Vector3(creature.GetPositionX(), creature.GetPositionY(), creature.GetPositionZ()));
+    Vector3 startPos, endPos;
+    uint32 travelTime = 0;
+    uint32 nextDelay = 0;
+    float nextOrientation = 0.f;
+    uint32 loops = 0;
+
+    do
     {
-        PointsArray genPath;
-        genPath.resize(pSubpath->size());
-        for (auto const& node : *pSubpath)
-            genPath[node.first] = G3D::Vector3(node.second.x, node.second.y, node.second.z);
+        WaypointNode const& currNode = currPoint->second;
+        nextDelay = currNode.delay;
+        nextOrientation = currNode.orientation;
+        startPos = genPath[genPath.size() - 1];
+        endPos = Vector3(currNode.x, currNode.y, currNode.z);
 
-        if (creature.CanFly())
-            init.SetFly();
+        if (WaypointPath const* pSubpath = (currNode.path_id ? sWaypointMgr.GetPathFromOrigin(currNode.path_id, 0, 0, PATH_FROM_SPECIAL) : nullptr))
+        {
+            PointsArray tmp;
+            tmp.push_back(startPos);
+            for (auto const& node : *pSubpath)
+                tmp.push_back(Vector3(node.second.x, node.second.y, node.second.z));
 
-        init.MovebyPath(genPath);
-    }
-    else
-        init.MoveTo(nextNode.x, nextNode.y, nextNode.z, (m_PathOrigin == PATH_FROM_SPECIAL) ? MOVE_STRAIGHT_PATH : MOVE_PATHFINDING);
+            travelTime += BuildIntPath(genPath, tmp) / m_creatureSpeed * 1000.f;
+
+            if (creature.CanFly())
+                init.SetFly();
+        }
+        else if (m_PathOrigin != PATH_FROM_SPECIAL)
+        {
+            PathFinder pathfinder(&creature);
+            pathfinder.calculate(startPos, endPos, true);
+            travelTime += BuildIntPath(genPath, pathfinder.getPath()) / m_creatureSpeed * 1000.f;
+        }
+        else
+        {
+            PointsArray tmp;
+            tmp.push_back(startPos);
+            tmp.push_back(endPos);
+            travelTime += BuildIntPath(genPath, tmp) / m_creatureSpeed * 1000.f;
+        }
 
-    if (nextNode.orientation != 100 && nextNode.delay != 0)
-        init.SetFacing(nextNode.orientation);
+        m_nodeIndexes.push_back(genPath.size() - 1);
+
+        if (nextDelay)
+            break;
+        else
+        {
+            ++currPoint;
+            if (currPoint == i_path->end())
+            {
+                if (!m_repeating)
+                    break;
+                else
+                {
+                    currPoint = i_path->begin();
+                    ++loops;
+                }
+            }
+        }
```

Comment:
If you avoid else after breaks, you can make the code a bit cleaner to look at.
```suggestion
        if (nextDelay)
            break;

        ++currPoint;

        if (currPoint == i_path->end()) 
        {
            if (!m_repeating)
                break;

            currPoint = i_path->begin();
            ++loops;
        }
```